pensieve/statistics.py

```diff
@@ -0,0 +1,158 @@
+            bigquery.SchemaField("parameter", "NUMERIC"),
+            bigquery.SchemaField("label", "STRING"),
+            bigquery.SchemaField("ci_width", "FLOAT64"),
+            bigquery.SchemaField("point", "FLOAT64"),
+            bigquery.SchemaField("lower", "FLOAT64"),
+            bigquery.SchemaField("upper", "FLOAT64"),
+        ]
+
+        if append:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_APPEND
+        else:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_TRUNCATE
+
+        client.load_table_from_json(
+            self.to_dict()["data"], destination_table, job_config=job_config
+        )
+
+    def to_dict(self):
+        """Return statistic results as dict."""
+
+        return attr.asdict(self)
+
+
+@attr.s(auto_attribs=True)
+class Statistic(ABC):
+    """
+    Abstract representation of a statistic.
+
+    A statistic is a transformation that accepts a table of per-client aggregates and
+    returns a table representing a summary of the aggregates with respect to the branches
+    of the experiment.
+    """
+
+    pre_treatments: List[PreTreatment]
+
+    @classmethod
+    def name(cls):
+        """Return snake-cased name of the statistic."""
+        return re.sub(r"(?<!^)(?=[A-Z])", "_", cls.__name__).lower()
+
+    def apply(self, df: DataFrame, metric: str) -> "StatisticResultCollection":
+        """
+        Run statistic on data provided by a DataFrame and return a collection
+        of statistic results.
+        """
+
+        statistic_result_collection = StatisticResultCollection([])
+
+        if metric in df:
+            data = df
+            for pre_treatment in self.pre_treatments:
+                data = pre_treatment.apply(data)
+
+            statistic_result_collection.data += self.transform(df, metric).data
+
+        return statistic_result_collection
+
+    @abstractmethod
+    def transform(self, df: DataFrame, metric: str) -> "StatisticResultCollection":
+        return NotImplemented
+
+    @classmethod
+    def from_config(cls, config_dict: Dict[str, Any]):  # todo: plug in config file support
+        """Create a class instance with the specified config parameters."""
+        return cls(**config_dict)
+
+
+@attr.s(auto_attribs=True)
+class BootstrapMean(Statistic):
+    num_samples: int = 100
+    confidence_interval: float = 0.95
+    pre_treatments: List[PreTreatment] = [RemoveNulls()]
+    branches: List[str] = []
+
+    def transform(self, df: DataFrame, metric: str) -> "StatisticResultCollection":
+        stats_results = StatisticResultCollection([])
+
+        results_per_branch = df.groupby("branch")
+
+        critical_point = (1 - self.confidence_interval) / 2
+        summary_quantiles = (critical_point, 1 - critical_point)
+
+        for branch in self.branches:
+            branch_data = results_per_branch.get_group(branch)
+            ma_result = mabsbb.bootstrap_one_branch(
```

Comment:
We should probably use the compare_branches version of this instead and let it handle discovering the branches.